brave/brave-core
# WIP: Add funds to wallet via Uphold popup

Fixes brave/brave-browser#923

* Added a new brave rewards web ui call to show Uphold poup when "Add Funds" button is clicked;
* Added AddFundsPopup class that manages Uphold popup window and modifies shields permissions for relevant sites while the popup is shown;
* Extended BraveLocationBarView with a call to hide brave actions container;
* Added a browser test to verify shields permissions changes before and after popup;
* Added a browser test to verify that popup is closed when brave://rewards is closed;
* Added a build flag to control Add Funds popup's availability in the build;
* Added a new brave rewads web ui event to notify the js ui when the popup's availability;
* Added js tests for the add funds call and popup availability event.

## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on

## Test Plan:

**Automated tests:**
`npm run test -- brave_browser_tests --filter=BraveAddFundsPopupTest.TestAddFundsPopupClosed`
`npm run test -- brave_browser_tests --filter=BraveAddFundsPopupTest.TestAddFundsPopupDeleted`

**Manual tests:**

1. **Test that popup shows:**
- Open Brave browser and navigate to brave://rewards,
- Click "Add Funds" button in "Your wallet" section,
- Verify that a popup with Uphold integration shows,
- Verify that the location bar in the popup does not show Brave shields or rewards action buttons,
- Test Uphold integration flows (shown in uphold-pay-flows.pdf).

2. **Test that popup is brought to focus when already open:**
- Open Brave browser and navigate to brave://rewards,
- Click "Add Funds" button in "Your wallet" section,
- Verify that a popup with Uphold integration shows,
- Bring the main browser window to the front by clicking on its title bar,
- Click "Add Funds" button in "Your wallet" section again,
- Verify that the previously opened popup is brought to the front and there's no new popup.

3. **Test that popup is closed when brave://rewards is closed:**
- Open Brave browser and navigate to brave://rewards,
- Click "Add Funds" button in "Your wallet" section,
- Verify that a popup with Uphold integration shows,
- Bring the main browser window to the front by clicking on its title bar,
- Open a new tab in the main window,
- Close the tab that contains brave://rewards,
- Verify that the popup is also closed.

4. **Test that default shields settings are restored after popup closes:**
- Start with a fresh profile and open Brave browser,
- Navigate to https://uphold-widget.herokuapp.com/ and open Shields settings for the site (note the settings for cookies and device recognition - should be set to block 3-rd party),
- In a new tab, navigate to brave://settings/content/autoplay and verify that the allow list is empty;
- Navigate to brave://settings/content/camera and verify that the allow list is empty,
- In a new tab, navigate to brave://rewards and click "Add Funds" button in "Your wallet" section,
- In the tab that has https://uphold-widget.herokuapp.com/ open Shields settings and verify that the settings for cookies and device recognition are set to allowed,
- In brave://settings/content/autoplay verify that https://uphold-widget.herokuapp.com:443 is listed in the allowed list,
- In brave://settings/content/camera verify that https://uphold-widget.herokuapp.com:443 is listed in the allowed list,
- Close uphold popup,
- In the tab that has https://uphold-widget.herokuapp.com/ open Shields settings and verify that the settings for cookies and device recognition are reset to the values they had before the popup was shown,
- In brave://settings/content/autoplay verify that https://uphold-widget.herokuapp.com:443 is not listed in the allowed list,
- In brave://settings/content/camera verify that https://uphold-widget.herokuapp.com:443 is not listed in the allowed list.

5. **Test that custom shields settings are restored after popup closes:**
Repeat previous test with the following modifications:
- Before invoking the uphold popup, modify shields settings (for example: set everything to block),
- Invoke the popup and verify that the expected in the previous test settings are set to allowed,
- Close the popup and verify that the settings are reset back to the settings before the popup was shown.

## Reviewer Checklist:
